diff -Nur hostapd20-00/hostapd/Makefile hostapd20/hostapd/Makefile
--- hostapd20-00/hostapd/Makefile	2013-01-12 16:42:54.000000000 +0200
+++ hostapd20/hostapd/Makefile	2013-04-23 15:44:25.000000000 +0300
@@ -1,7 +1,3 @@
-ifndef CC
-CC=gcc
-endif
-
 ifndef CFLAGS
 CFLAGS = -MMD -O2 -Wall -g
 endif
@@ -100,7 +96,8 @@
 ifndef CONFIG_NO_DUMP_STATE
 # define HOSTAPD_DUMP_STATE to include SIGUSR1 handler for dumping state to
 # a file (undefine it, if you want to save in binary size)
-CFLAGS += -DHOSTAPD_DUMP_STATE
+CFLAGS += -DHOSTAPD_DUMP_STATE $(IFX_CFLAGS)
+LDFLAGS = $(IFX_LDFLAGS)
 OBJS += dump_state.o
 OBJS += ../src/eapol_auth/eapol_auth_dump.o
 endif
@@ -854,7 +851,8 @@
 BCHECK=../src/drivers/build.hostapd
 
 hostapd: $(BCHECK) $(OBJS)
-	$(Q)$(CC) $(LDFLAGS) -o hostapd $(OBJS) $(LIBS)
+	$(Q)$(CC) $(LDFLAGS) -o hostapd $(OBJS) $(LIBS) $(LDFLAGS)
+	if [ "$(STRIP)" ]; then cp hostapd hostapd_debug && $(STRIP) hostapd; fi
 	@$(E) "  LD " $@
 
 ifdef CONFIG_WPA_TRACE
diff -Nur hostapd20-00/wpa_supplicant/Makefile hostapd20/wpa_supplicant/Makefile
--- hostapd20-00/wpa_supplicant/Makefile	2013-01-12 16:42:54.000000000 +0200
+++ hostapd20/wpa_supplicant/Makefile	2013-04-23 15:44:25.000000000 +0300
@@ -6,10 +6,13 @@
 CFLAGS = -MMD -O2 -Wall -g
 endif
 
-export LIBDIR ?= /usr/local/lib/
-export BINDIR ?= /usr/local/sbin/
 PKG_CONFIG ?= pkg-config
 
+# Include directories for CVS version
+#CFLAGS += -I. -I../utils -I../hostapd
+LFLAGS = $(IFX_LDFLAGS)
+CFLAGS += $(IFX_CFLAGS)
+
 CFLAGS += -I../src
 CFLAGS += -I../src/utils
 
@@ -1507,23 +1510,26 @@
 $(OBJS_c) $(OBJS_t) $(OBJS_t2) $(OBJS) $(BCHECK) $(EXTRA_progs): .config
 
 wpa_supplicant: $(BCHECK) $(OBJS) $(EXTRA_progs)
-	$(Q)$(LDO) $(LDFLAGS) -o wpa_supplicant $(OBJS) $(LIBS) $(EXTRALIBS)
+	$(Q)$(LDO) $(LDFLAGS) -o wpa_supplicant $(OBJS) $(LIBS) $(EXTRALIBS) $(LFLAGS)
+	if [ "$(STRIP)" ]; then cp wpa_supplicant wpa_supplicant_debug && $(STRIP) wpa_supplicant; fi
 	@$(E) "  LD " $@
 
 eapol_test: $(OBJS_t)
-	$(Q)$(LDO) $(LDFLAGS) -o eapol_test $(OBJS_t) $(LIBS)
+	$(Q)$(LDO) $(LDFLAGS) -o eapol_test $(OBJS_t) $(LIBS) $(LFLAGS)
 	@$(E) "  LD " $@
 
 preauth_test: $(OBJS_t2)
-	$(Q)$(LDO) $(LDFLAGS) -o preauth_test $(OBJS_t2) $(LIBS)
+	$(Q)$(LDO) $(LDFLAGS) -o preauth_test $(OBJS_t2) $(LIBS) $(LFLAGS)
 	@$(E) "  LD " $@
 
 wpa_passphrase: $(OBJS_p)
-	$(Q)$(LDO) $(LDFLAGS) -o wpa_passphrase $(OBJS_p) $(LIBS_p)
+	$(Q)$(LDO) $(LDFLAGS) -o wpa_passphrase $(OBJS_p) $(LIBS_p) $(LFLAGS)
+	if [ "$(STRIP)" ]; then cp wpa_passphrase wpa_passphrase_debug && $(STRIP) wpa_passphrase; fi
 	@$(E) "  LD " $@
 
 wpa_cli: $(OBJS_c)
-	$(Q)$(LDO) $(LDFLAGS) -o wpa_cli $(OBJS_c) $(LIBS_c)
+	$(Q)$(LDO) $(LDFLAGS) -o wpa_cli $(OBJS_c) $(LIBS_c) $(LFLAGS)
+	if [ "$(STRIP)" ]; then cp wpa_cli wpa_cli_debug && $(STRIP) wpa_cli; fi
 	@$(E) "  LD " $@
 
 link_test: $(OBJS) $(OBJS_h) tests/link_test.o
