--- /dev/null
+++ b/ChangeLog
@@ -0,0 +1,37 @@
+
+2014-01-26:
+	Skipping some code if CONFIG_WPS2 defined
+
+2013-11-27:
+	10_hostapd_enable_background_task_console.patch added
+
+2013-06-09:
+	mtlk_deinit support added
+
+2013-05-19:
+	Added wrapers around ioctl 
+
+2013-04-24:
+	wpa_printf used instead of printf
+
+2013-04-24:
+	Enabled CONFIG_IEEE80211W
+
+2013-04-24:
+	Support added for loglevels
+
+2013-04-24:
+	06_hostapd_probe_req_dbg_cleanup.patch added
+
+2013-04-23:
+	Increased essis.length by 1
+
+2013-04-23:
+	configuration options for building hostapd
+
+2013-04-23:
+	Support for communication between the driver and userspace apps using generic netlink library
+
+2013-04-23:
+	Cflag changes for compilation
+
