diff --exclude 'msvc*' -Npur orig/src/drivers/driver_nl80211.c patched/src/drivers/driver_nl80211.c
--- orig/src/drivers/driver_nl80211.c	2015-10-12 09:08:13.000000000 +0300
+++ patched/src/drivers/driver_nl80211.c	2015-10-15 13:06:41.000000000 +0300
@@ -5228,7 +5228,7 @@ static int wpa_driver_nl80211_scan(struc
 	drv->scan_state = SCAN_REQUESTED;
 	/* Not all drivers generate "scan completed" wireless event, so try to
 	 * read results after a timeout. */
-	timeout = 10;
+	timeout = 30;
 	if (drv->scan_complete_events) {
 		/*
 		 * The driver seems to deliver events to notify when scan is
