diff --exclude 'msvc*' -Npur orig/src/ap/ieee802_11.c patched/src/ap/ieee802_11.c
--- orig/src/ap/ieee802_11.c	2015-10-12 09:08:13.000000000 +0300
+++ patched/src/ap/ieee802_11.c	2015-10-13 09:51:26.000000000 +0300
@@ -774,16 +774,7 @@ static void ap_sta_remove_in_other_bss_n
 		wpa_printf(MSG_DEBUG, "Removing station " MACSTR
 				   " with AID=%d from kernel driver.", MAC2STR(sta2->addr), sta2->aid);
 
-		if (hostapd_drv_sta_remove(bss, sta2->addr))
-			wpa_printf(MSG_DEBUG, "Could not remove station " MACSTR
-					   " from kernel driver.", MAC2STR(sta2->addr));
-
-		if (sta2->aid > 0) {
-			if (bss->driver->free_aid)
-				(*bss->driver->free_aid)(bss->drv_priv, &sta2->aid);
-		}
-		else
-			wpa_printf(MSG_DEBUG, "Can't remove AID, (AID=%d)", sta2->aid);
+		ap_free_sta(bss, sta2);
 	}
 }
 
diff --exclude 'msvc*' -Npur orig/src/ap/sta_info.c patched/src/ap/sta_info.c
--- orig/src/ap/sta_info.c	2015-10-12 09:08:13.000000000 +0300
+++ patched/src/ap/sta_info.c	2015-10-13 09:53:15.000000000 +0300
@@ -148,6 +148,7 @@ void ap_free_sta(struct hostapd_data *ha
 {
 	int set_beacon = 0;
 
+	wpa_printf(MSG_MSGDUMP, "ap_free_sta");
 	accounting_sta_stop(hapd, sta);
 
 	/* just in case */
