diff -Npur hostap-724fce8-orig/hostapd/main.c hostap-724fce8-modif/hostapd/main.c
--- hostap-724fce8-orig/hostapd/main.c	2016-02-11 14:23:41.000000000 +0200
+++ hostap-724fce8-modif/hostapd/main.c	2016-02-11 14:45:50.000000000 +0200
@@ -720,8 +720,6 @@ int main(int argc, char *argv[])
 
 	hostapd_global_ctrl_iface_init(&interfaces);
 
-	hostapd_dfs_debug_channel_init(interfaces.iface[0]);
-
 	if (hostapd_global_run(&interfaces, daemonize, pid_file)) {
 		wpa_printf(MSG_ERROR, "Failed to start eloop");
 		goto out;
diff -Npur hostap-724fce8-orig/src/ap/dfs.c hostap-724fce8-modif/src/ap/dfs.c
--- hostap-724fce8-orig/src/ap/dfs.c	2016-02-11 14:23:42.000000000 +0200
+++ hostap-724fce8-modif/src/ap/dfs.c	2016-02-11 14:45:50.000000000 +0200
@@ -989,7 +989,7 @@ switch_chan:
 	return 0;
 }
 
-void hostapd_dfs_debug_channel_init(struct hostapd_iface *iface)
+void hostapd_send_dfs_debug_channel_flag(struct hostapd_iface *iface)
 {
 	int i;
 	u8 dfs_required = 0;
@@ -998,6 +998,10 @@ void hostapd_dfs_debug_channel_init(stru
 	struct hostapd_data *hapd = iface->bss[0];
 	struct hostapd_hw_modes *mode = iface->current_mode;
 
+	if (!mode || !hapd) {
+		wpa_printf(MSG_ERROR, "HW mode or hostapd data not initialized, dfs debug channel flag won't be sent");
+		return;
+	}
 	
 	if (iface->conf->dfs_debug_chan) {
 		for (i = 0; i < mode->num_channels; i++) {
@@ -1012,7 +1016,7 @@ void hostapd_dfs_debug_channel_init(stru
 	}
 	
 	if (hostapd_drv_vendor_cmd(hapd, OUI_LTQ, LTQ_NL80211_VENDOR_SUBCMD_DFS_DEBUG,
-														 &dfs_required, sizeof(dfs_required), NULL))
+							   &dfs_required, sizeof(dfs_required), NULL))
 		wpa_printf(MSG_INFO, "Failed to send driver vendor command LTQ_NL80211_VENDOR_SUBCMD_DFS_DEBUG");
 }
 
diff -Npur hostap-724fce8-orig/src/ap/dfs.h hostap-724fce8-modif/src/ap/dfs.h
--- hostap-724fce8-orig/src/ap/dfs.h	2016-02-11 14:23:42.000000000 +0200
+++ hostap-724fce8-modif/src/ap/dfs.h	2016-02-11 14:45:50.000000000 +0200
@@ -22,6 +22,6 @@ int hostapd_dfs_nop_finished(struct host
 			     int ht_enabled,
 			     int chan_offset, int chan_width, int cf1, int cf2);
 int hostapd_is_dfs_required(struct hostapd_iface *iface);
-void hostapd_dfs_debug_channel_init(struct hostapd_iface *iface);
+void hostapd_send_dfs_debug_channel_flag(struct hostapd_iface *iface);
 
 #endif /* DFS_H */
diff -Npur hostap-724fce8-orig/src/ap/hostapd.c hostap-724fce8-modif/src/ap/hostapd.c
--- hostap-724fce8-orig/src/ap/hostapd.c	2016-02-11 14:23:42.000000000 +0200
+++ hostap-724fce8-modif/src/ap/hostapd.c	2016-02-11 14:45:50.000000000 +0200
@@ -1153,6 +1153,7 @@ static int setup_interface2(struct hosta
 				   "channel. (%d)", ret);
 			goto fail;
 		}
+		hostapd_send_dfs_debug_channel_flag(iface);
 		if (ret == 1) {
 			wpa_printf(MSG_DEBUG, "Interface initialization will be completed in a callback (ACS)");
 			return 0;
