diff --git a/include/linux/if.h b/include/linux/if.h
index 722ad71..184784a 100644
--- a/include/linux/if.h
+++ b/include/linux/if.h
@@ -53,6 +53,9 @@
 
 #define IFF_ECHO	0x40000		/* echo sent packets		*/
 
+#define IFF_NOMULTIPATH	0x80000		/* Disable for MPTCP */
+#define IFF_MPMASTER    0x200000	/* Use as a master path for MPTCP */
+
 #define IFF_VOLATILE	(IFF_LOOPBACK|IFF_POINTOPOINT|IFF_BROADCAST|IFF_ECHO|\
 		IFF_MASTER|IFF_SLAVE|IFF_RUNNING|IFF_LOWER_UP|IFF_DORMANT)
 
diff --git a/ip/ipaddress.c b/ip/ipaddress.c
index 85f05a2..29d017f 100644
--- a/ip/ipaddress.c
+++ b/ip/ipaddress.c
@@ -95,6 +95,8 @@ void print_link_flags(FILE *fp, unsigned flags, unsigned mdown)
 	_PF(BROADCAST);
 	_PF(POINTOPOINT);
 	_PF(MULTICAST);
+	_PF(NOMULTIPATH);
+	_PF(MPMASTER);
 	_PF(NOARP);
 	_PF(ALLMULTI);
 	_PF(PROMISC);
diff --git a/ip/iplink.c b/ip/iplink.c
index 48c0254..65b2330 100644
--- a/ip/iplink.c
+++ b/ip/iplink.c
@@ -59,6 +59,8 @@ void iplink_usage(void)
 	fprintf(stderr, "	                  [ dynamic { on | off } ]\n");
 	fprintf(stderr, "	                  [ multicast { on | off } ]\n");
 	fprintf(stderr, "	                  [ allmulticast { on | off } ]\n");
+	fprintf(stderr, "	                  [ multipath { on | off } ]\n");
+	fprintf(stderr, "	                  [ multipath { on | off | backup | master | nomaster } ]\n");
 	fprintf(stderr, "	                  [ promisc { on | off } ]\n");
 	fprintf(stderr, "	                  [ trailers { on | off } ]\n");
 	fprintf(stderr, "	                  [ txqueuelen PACKETS ]\n");
@@ -325,6 +327,20 @@ int iplink_parse(int argc, char **argv, struct iplink_req *req,
 				req->i.ifi_flags &= ~IFF_ALLMULTI;
 			} else
 				return on_off("allmulticast");
+		} else if (strcmp(*argv, "multipath") == 0) {
+			NEXT_ARG();
+			req->i.ifi_change |= IFF_NOMULTIPATH;
+			req->i.ifi_change |= IFF_MPMASTER;
+			if (strcmp(*argv, "on") == 0) {
+				req->i.ifi_flags &= ~IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "off") == 0) {
+				req->i.ifi_flags |= IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "master") == 0) {
+				req->i.ifi_flags |= IFF_MPMASTER;
+			} else if (strcmp(*argv, "nomaster") == 0) {
+				req->i.ifi_flags &= ~IFF_MPMASTER;
+			} else
+				return on_off("multipath");
 		} else if (strcmp(*argv, "promisc") == 0) {
 			NEXT_ARG();
 			req->i.ifi_change |= IFF_PROMISC;
@@ -798,6 +814,20 @@ static int do_set(int argc, char **argv)
 				flags &= ~IFF_ALLMULTI;
 			} else
 				return on_off("allmulticast");
+		} else if (strcmp(*argv, "multipath") == 0) {
+			NEXT_ARG();
+			mask |= IFF_NOMULTIPATH;
+			mask |= IFF_MPMASTER;
+			if (strcmp(*argv, "on") == 0) {
+				flags &= ~IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "off") == 0) {
+				flags |= IFF_NOMULTIPATH;
+			} else if (strcmp(*argv, "master") == 0) {
+				flags |= IFF_MPMASTER;
+			} else if (strcmp(*argv, "nomaster") == 0) {
+				flags &= ~IFF_MPMASTER;
+			} else
+				return on_off("multipath");
 		} else if (strcmp(*argv, "promisc") == 0) {
 			NEXT_ARG();
 			mask |= IFF_PROMISC;
diff --git a/man/man8/ip.8 b/man/man8/ip.8
index c5248ef..b4b0185 100644
--- a/man/man8/ip.8
+++ b/man/man8/ip.8
@@ -68,6 +68,8 @@ ip \- show / manipulate routing, devices, policy routing and tunnels
 .br
 .BR multicast " { " on " | " off " } |"
 .br
+.BR multipath " { " on " | " off " | " master " | " nomaster " } |"
+.br
 .B  txqueuelen
 .IR PACKETS " |"
 .br
@@ -963,6 +965,12 @@ change the
 flag on the device.
 
 .TP
+.BR "multipath on " or " multipath off"
+change the
+.B MULTIPATH
+flag on the device, to enable or disable MPTCP.
+
+.TP
 .BR "dynamic on " or " dynamic off"
 change the
 .B DYNAMIC
